Lib/asyncio/streams.py

```diff
@@ -91,40 +222,301 @@ def factory():
     return await loop.create_server(factory, host, port, **kwds)
 
 
+class _BaseStreamServer:
+    # Design notes.
+    # StreamServer and UnixStreamServer are exposed as FINAL classes,
+    # not function factories.
+    # async with serve(host, port) as server:
+    #      server.start_serving()
+    # looks ugly.
+    # The class doesn't provide API for enumerating connected streams
+    # It can be a subject for improvements in Python 3.9
+
+    _server_impl = None
+
+    def __init__(self, client_connected_cb,
+                 /,
+                 limit=_DEFAULT_LIMIT,
+                 shutdown_timeout=60,
+                 _asyncio_internal=False):
+        if not _asyncio_internal:
+            raise RuntimeError("_ServerStream is a private asyncio class")
+        self._client_connected_cb = client_connected_cb
+        self._limit = limit
+        self._loop = events.get_running_loop()
+        self._streams = {}
+        self._shutdown_timeout = shutdown_timeout
+
+    def __init_subclass__(cls):
+        if not cls.__module__.startswith('asyncio.'):
+            raise TypeError(f"asyncio.{cls.__name__} "
+                            "class cannot be inherited from")
+
+    async def bind(self):
+        if self._server_impl is not None:
+            return
+        self._server_impl = await self._bind()
+
+    def is_bound(self):
+        return self._server_impl is not None
+
+    @property
+    def sockets(self):
+        # multiple value for socket bound to both IPv4 and IPv6 families
+        if self._server_impl is None:
+            return []
+        return self._server_impl.sockets
+
+    def is_serving(self):
+        if self._server_impl is None:
+            return False
+        return self._server_impl.is_serving()
+
+    async def start_serving(self):
+        await self.bind()
+        await self._server_impl.start_serving()
+
+    async def serve_forever(self):
+        await self.start_serving()
+        await self._server_impl.serve_forever()
+
+    async def close(self):
+        if self._server_impl is None:
```

Comment:
This isn't correct.  Because we use `Stream` in subprocess APIs, this is not a backwards compatible change.  Please make it synchronous and you can return an `_OptionalAwait` from it.

Reply:
My bad, this is a different class :)